pkg/server/service/proxy.go

```diff
@@ -37,45 +38,71 @@ func buildProxy(passHostHeader *bool, responseForwarding *dynamic.ResponseForwar
 	}
 
 	proxy := &httputil.ReverseProxy{
-		Director: func(outReq *http.Request) {
-			u := outReq.URL
-			if outReq.RequestURI != "" {
-				parsedURL, err := url.ParseRequestURI(outReq.RequestURI)
+		Rewrite: func(req *httputil.ProxyRequest) {
+			u := req.In.URL
+			if req.In.RequestURI != "" {
+				parsedURL, err := url.ParseRequestURI(req.In.RequestURI)
 				if err == nil {
 					u = parsedURL
 				}
 			}
 
-			outReq.URL.Path = u.Path
-			outReq.URL.RawPath = u.RawPath
+			req.Out.URL.Path = u.Path
+			req.Out.URL.RawPath = u.RawPath
 			// If a plugin/middleware adds semicolons in query params, they should be urlEncoded.
-			outReq.URL.RawQuery = strings.ReplaceAll(u.RawQuery, ";", "&")
-			outReq.RequestURI = "" // Outgoing request should not have RequestURI
+			req.Out.URL.RawQuery = strings.ReplaceAll(u.RawQuery, ";", "&")
+			req.Out.RequestURI = "" // Outgoing request should not have RequestURI
 
-			outReq.Proto = "HTTP/1.1"
-			outReq.ProtoMajor = 1
-			outReq.ProtoMinor = 1
-
-			// Do not pass client Host header unless optsetter PassHostHeader is set.
+			// Do not pass client Host header unless PassHostHeader is set.
 			if passHostHeader != nil && !*passHostHeader {
-				outReq.Host = outReq.URL.Host
+				req.Out.Host = req.In.URL.Host
+			}
+
+			// Add back removed Forwarded Headers.
+			req.Out.Header["Forwarded"] = req.In.Header["Forwarded"]
+			req.Out.Header["X-Forwarded-For"] = req.In.Header["X-Forwarded-For"]
+			req.Out.Header["X-Forwarded-Host"] = req.In.Header["X-Forwarded-Host"]
+			req.Out.Header["X-Forwarded-Proto"] = req.In.Header["X-Forwarded-Proto"]
+
+			remoteAddr := req.In.RemoteAddr
+			// In case of a ProxyProtocol connection the http.Request#RemoteAddr is the original one,
+			// and if the Forwarded headers were trusted from the peer socket address (see trustedIPs) and not from the original IP,
+			// the peer socket address should be used instead.
+			// The ForwardedHeaders middleware provides the address to be used to populate the X-Forwarded-For header.
+			if xForwardedForAddr, ok := req.In.Context().Value(forwardedheaders.XForwardedForAddr).(string); ok {
+				remoteAddr = xForwardedForAddr
+			}
+
+			// Adapted from httputil.ReverseProxy.
+			if clientIP, _, err := net.SplitHostPort(remoteAddr); err == nil {
```

Comment:
I think that when the port is provided in the Proxy Protocol, we should populate the ```X-Forwarded-Port``` header with that value, rather than ignoring it.